dids/src/main/kotlin/web5/sdk/dids/didcore/DIDDocument.kt

```diff
@@ -0,0 +1,295 @@
+      ?: throw Exception("No verification method found for id: $vmID")
+    return vm
+  }
+
+  /**
+   * GetAbsoluteResourceID returns a fully qualified ID for a document resource (e.g. service, verification method)
+   * Document Resource IDs are allowed to be relative DID URLs as a means to reduce storage size of DID Documents.
+   * More info here: https://www.w3.org/TR/did-core/#relative-did-urls
+   *
+   * @param id of the resource
+   * @return fully qualified ID for a document resource
+   */
+  public fun getAbsoluteResourceID(id: String): String {
+    return if (id.startsWith("#")) "${this.id}$id" else id
+  }
+
+  /**
+   * Finds the first available assertion method from the [DIDDocument]. When [assertionMethodId]
+   * is null, the function will return the first available assertion method.
+   *
+   * @param assertionMethodId The ID of the assertion method to be found
+   * @return VerificationMethod with purpose of Assertion
+   */
+  @JvmOverloads
+  public fun findAssertionMethodById(assertionMethodId: String? = null): VerificationMethod {
+    require(!assertionMethod.isNullOrEmpty()) {
+      throw SignatureException("No assertion methods found in DID document")
+    }
+
+    if (assertionMethodId != null) {
+      require(assertionMethod.contains(assertionMethodId)) {
+        throw SignatureException("assertion method \"$assertionMethodId\" not found in list of assertion methods")
+      }
+    }
+
+    val assertionMethod: VerificationMethod =
+      verificationMethod
+        ?.find {
+          it.id == (assertionMethodId ?: assertionMethod.first())
+        }
+        ?: throw SignatureException("assertion method \"$assertionMethodId\" not found")
+
+    return assertionMethod
+  }
+
+  /**
+   * Builder object to build a DIDDocument.
+   */
+  public class Builder {
+
+    private var id: String? = null
+    private var context: String? = null
+    private var alsoKnownAs: List<String>? = null
+    private var controller: List<String>? = null
+
+    private var verificationMethod: MutableSet<VerificationMethod>? = null
+    private var service: List<Service>? = null
+
+    private var assertionMethod: MutableList<String>? = null
+    private var authenticationMethod: MutableList<String>? = null
+    private var keyAgreementMethod: MutableList<String>? = null
+    private var capabilityDelegationMethod: MutableList<String>? = null
+    private var capabilityInvocationMethod: MutableList<String>? = null
+
+    /**
+     * Adds Id to the DIDDocument.
+     *
+     * @param id of the DIDDocument
+     * @return Builder object
+     */
+    public fun id(id: String): Builder = apply { this.id = id }
+
+    /**
+     * Adds Context to the DIDDocument.
+     *
+     * @param context of the DIDDocument
+     * @return Builder object
+     */
+    public fun context(context: String): Builder = apply {
+      this.context = context
+    }
+
+    /**
+     * Adds Controllers.
```

Comment:
Might be better to use the word "Sets" rather than "Adds" as with arrays/lists "Add" may be confused with append or concat